blockdata/constraints.lisp

```diff
@@ -1,147 +1,380 @@
-
-(defconstraint value-constraints---GASLIMIT (:guard (blockdata---first-row-of-new-block))
-               (begin  
-                 (eq!       (shift INST      ROW_SHIFT_GASLIMIT) EVM_INST_GASLIMIT)
-                 (vanishes! (shift DATA_HI   ROW_SHIFT_GASLIMIT))
-                 (eq!       (shift DATA_LO   ROW_SHIFT_GASLIMIT) LINEA_BLOCK_GAS_LIMIT)
-                 (eq!       (shift DATA_LO   ROW_SHIFT_GASLIMIT) BLOCK_GAS_LIMIT)))
-
-(defconstraint value-constraints---CHAINID (:guard (blockdata---first-row-of-new-block))
-               (begin  
-                 (eq!       (shift INST      ROW_SHIFT_CHAINID) EVM_INST_CHAINID)
-                 (vanishes! (shift DATA_HI   ROW_SHIFT_CHAINID))
-                 ;(eq!      (shift DATA_LO   ROW_SHIFT_CHAINID) LINEA_CHAIN_ID) ;; TODO: this needs some fixing
-                 ))
-
-(defconstraint value-constraints---BASEFEE (:guard (blockdata---first-row-of-new-block))
-               (begin  
-                 (eq!       (shift INST      ROW_SHIFT_BASEFEE) EVM_INST_BASEFEE)
-                 (vanishes! (shift DATA_HI   ROW_SHIFT_BASEFEE))
-                 (eq!       (shift DATA_LO   ROW_SHIFT_BASEFEE) LINEA_BASE_FEE)
-                 (eq!       (shift DATA_LO   ROW_SHIFT_BASEFEE) BASEFEE)))
+         (counter-constancy CT (wght-sum))))
+
+(defconstraint first-block-number-is-conflation-constant ()
+  (if-not-zero IOMF
+      (if-not-zero (shift IOMF 1)
+           (eq! (shift FIRST_BLOCK_NUMBER 1) FIRST_BLOCK_NUMBER))))
+
+;;;;;;;;;;;;;;;;;;;;;
+;;                 ;;
+;;  2.5 Heartbeat  ;;
+;;                 ;;
+;;;;;;;;;;;;;;;;;;;;;
+
+(defconstraint first-row-iomf (:domain {0})
+  (vanishes! IOMF))
+
+(defconstraint padding-vanishing ()
+  (if-zero IOMF
+      (vanishes! CT)
+      (vanishes! (shift CT 1))))
+
+(defconstraint first-instruction-is-coinbase ()
+  (if-zero IOMF
+      (if-not-zero (shift IOMF 1)
+          (eq! (shift IS_CB 1) 1))))
+
+(defconstraint counter-reset-for-phase-entry ()
+  (if-not-zero (phase-entry)
+      (vanishes! (shift CT 1))))
+
+(defconstraint counter-increase-or-instruction-transition ()
+  (if-not-zero IOMF
+      (if-not-zero (- CT CT_MAX)
+          (eq! (shift CT 1) (+ 1 CT)))
+          (eq! (allowable-transitions) 1)))
+
+(defconstraint first-row-rel-block (:domain {0})
+  (vanishes! REL_BLOCK))
+
+(defconstraint rel-block-constant-or-increment ()
+  (any! (remained-constant! REL_BLOCK) (did-inc! REL_BLOCK 1)))
+
+(defconstraint rel-block-increment-from-basefee-to-coinbase ()
+  (eq! (shift REL_BLOCK 1) (+ REL_BLOCK ( * IS_BF (shift IS_CB 1)))))
+
+(defconstraint next-is-curr ()
+  (if-not-zero IS_CURR
+      (eq! (shift IS_CURR 1) 1)))
+
+(defconstraint is-prev-after-padding ()
+  (if-zero IOMF
+      (if-not-zero (shift IOMF 1)
+          (eq! (shift IS_PREV 1) 1))))
+
+(defconstraint curr-prev-wght-sum-constancy ()
+  (if-not-zero IOMF
+      (if-eq (shift REL_BLOCK 1) REL_BLOCK
```

Comment:
This should be implemented as 
```lisp
(if-not-zero (will-remain-constant! REL_BLOCK) ...)
```
or so. In any case, it should be implemented with an `if-not-zero` constraint, not an equality constraint.

<img width="546" alt="image" src="https://github.com/user-attachments/assets/42472df8-b7df-4c9c-b51a-ec38f72dfa63">